### ISSUE TITLE
Use PathParser instead of building paths in the code

### ISSUE THREAD

**Issue by bartek-wesolowski** [rafaeltonholo/svg-to-compose#24]
The converter produces code like this (fragment taken from one of the samples)
```kotlin
 // M256 0 c4.6 0 9.2 1 13.4 2.9 l188.3 79.9 c22 9.3 38.4 31 38.3 57.2 -0.5 99.2 -41.3 280.7 -213.6 363.2 -16.7 8 -36.1 8 -52.8 0 C57.3 420.7 16.5 239.2 16 140 c-0.1 -26.2 16.3 -47.9 38.3 -57.2 L242.7 2.9 C246.8 1 251.4 0 256 0z m0 66.8 v378 C394 378 431.1 230.1 432 141.4 L256 66.8z
path(
    fill = SolidColor(Color(0xFF1E3050)),
) {
    // M 256 0
    moveTo(x = 256.0f, y = 0.0f)
    // c 4.6 0 9.2 1 13.4 2.9
    curveToRelative(
        dx1 = 4.6f,
        dy1 = 0.0f,
        dx2 = 9.2f,
        dy2 = 1.0f,
        dx3 = 13.4f,
        dy3 = 2.9f,
    )
    // other instructions here
}
  ```
  Creating a path in code makes it quite verbose to the point where the IDE becomes unresponsive for larger images.
  
You could consider using a [PathParser](https://developer.android.com/reference/kotlin/androidx/core/graphics/PathParser) instead to produce something like this instead
```kotlin
addPath(
  PathParser().parsePathString(
    "M256 0 c4.6 0 9.2 1 13.4 2.9 l188.3 79.9 c22 9.3 38.4 31 38.3 57.2 -0.5 99.2 -41.3 280.7 -213.6 363.2 -16.7 8 -36.1 8 -52.8 0 C57.3 420.7 16.5 239.2 16 140 c-0.1 -26.2 16.3 -47.9 38.3 -57.2 L242.7 2.9 C246.8 1 251.4 0 256 0z m0 66.8 v378 C394 378 431.1 230.1 432 141.4 L256 66.8z"
  ).toNodes()
)
```
The result would be much shorter.

**Comment by rafaeltonholo** [comment]
Hi @bartek-wesolowski, thanks for the suggestion! 

Sorry for the long text ahead, but to answer your question, I need to give you a bit of context that led me to create this CLI.

# TL;DR
The author explains the reasons for creating a command-line tool to convert SVG/AVGs to Jetpack Compose icons. The main goals are to:

1. Avoid redundant calculations on the user's device.
2. Optimize the icon before generating the code.
3. Eliminate the need for XML drawables.

The author argues that using a pre-processed file with drawing instructions is more efficient than parsing the SVG data on the fly. While the generated file might be larger, the calculations happen only once during the creation of the file, not on every app launch or on the first time of icon usage in the app.

The author also acknowledges potential drawbacks like large file sizes and limitations for very complex SVG/AVGs, but offers solutions like minification and future improvements. 

In the end, the author asks the OP to share more details about the problematic SVG that makes the IDE unresponsive for further analysis.

> TL;DR section generated by Gemini

# Explanation

Here are the 3 main reasons why I created this CLI:
1. Avoid unnecessary calculations.
2. Provide a way to optimize the icon before generating it.
3. Provide a way to avoid using XML drawables (avg) on Compose.

<details><summary>Let me try to explain the 1st and the 3rd topics in depth (click to expand)</summary>
<p>

## 1st Topic - Avoid unnecessary calculations if possible.
I'll explain more about this in the next section, but using XML files does some calculations that we could avoid if we created the `ImageVector` directly, and that would also be avoided by using `PathParser().parsePathString(str).toNodes()`, since we are already creating the `ImageVector`.

However, we need to understand what happens when you call `PathParser().parsePathString(str).toNodes()`.

1. The `parsePathString` function will iterate over the whole path string trying to understand when a draw command starts/ends and whether or not it closes the current draw path.
2. Whenever it detects a command, it creates a `FloatArray` that will hold the draw command arguments and when it reaches the size of the current draw command, it creates a `PathNode` object that will contain the draw command and clear this `FloatArray` so it can look for the next command.
3. And then it adds this object to the `Node: ArrayList<PathNodes>'. 

They do this in a very optimized way, but it is still a calculation you have to do.

The `.toNodes()` function just returns the `Nodes` property, so no computation or calculation is done there.

## 3rd topic - XML Drawables on Compose
Whenever we want to use an XML drawable on Compose, we need to use the `painterResource' method which, in the case of AVG, generates a `VectorPainter' containing the icon/image so that Compose can render it.

If we navigate through the `painterResource` implementation, trying to simplify the steps to create the icon, we could describe it as
1. It will load the XML file, read its contents and parse it to an `XmlResourceParser`.
2. Create an `ImageVector.Builder' based on the vector attributes, including some calculations to determine the size of the vector.
3. Iterate over each node of the AVG so it can create the `ImageVector.Builder` instructions to draw the icon/image
    1. During parsing, it needs to identify if it is a group, a path or a clip-path to create the nodes correctly.
    2. If it is a path, it will load all the parameters related to the path and create the `ImageVector` nodes using the `PathParser(). parsePathString(str).toNodes()`, behind the scenes.
4. Cache the loaded vector so it doesn't have to perform the same calculation again.
5. Convert the `ImageVector` into a `VectorPainter`.

They do this in a very optimized way, but in the end it is always:
1. Read the XML file, if it is not yet cached.
2. Create the `ImageVector.Builder`.
3. Create the drawing instructions
4. Convert the `ImageVector` to a `VectorPainter` (even if the `ImageVector` was memorized).

So I saw an opportunity to avoid some calculations by having the `ImageVector` with all the drawing instructions already done by a CLI.

</p>
</details> 

## Why not use `PathParser().parsePathString(str).toNodes()` instead?
In the end, the `PathParser().parsePathString(str).toNodes()` will create the same list of `PathNodes` that we created using this CLI, but the computation to create the nodes will be done on the user's device, while using this CLI, it will be done only once on your PC or CI or wherever you use it.

If you take a look at each path function that we use inside the `ImageVector.Builder` lambda function, you'll see that they just add the statement to the node list by using the correct `PathNode` subclass constructor, however, it doesn't read or do any kind of calculation since it's expecting the values to it.

For this reason, I would rather have a large file rather than leverage the computation of creating path statements on the user side.

For small icons, this calculation may be meaningless, but for larger icons, it could have some impact, depending on how complex your application is.

## Disadvantages of the current approach
As you mentioned, the current approach could potentially generate very long files. Using the sample file [Illustration.nonoptimized.kt](https://github.com/rafaeltonholo/svg-to-compose/blob/main/samples/Illustration.avg.nonoptimized.kt) as an example, we can see that it generates a file with 5k lines, but there is a way to create a smaller file.

By using the `--minified` flag, it will generate a file without any comments explaining the SVG/AVG paths and will put all parameters on one line. It will generate a file with about 1k lines from the same svg. However, some linters may complain about the file because it doesn't follow the Kotlin style guide.

There is also a possibility that for very complex and large files, the icon cannot be rendered or the application cannot be compiled due to a JVM limitation. The current version can't handle this yet, but this issue is being addressed in the [feat/handle-gradient-and-item-tags] branch (https://github.com/rafaeltonholo/svg-to-compose/tree/feat/handle-gradient-and-item-tags), and the fix is coming soon.

Regarding the point of:
> Creating a path in code makes it quite verbose to the point where the IDE becomes unresponsive for larger images.

Are you experiencing this problem? Would you mind sharing the SVG/AVG you are using so I can evaluate the situation on my end?

I've tested several complex SVGs, including large ones that broke the preview due to the JVM limitation, and I haven't experienced this, so it would be nice to have an SVG that causes this to better understand what's happening.

**Comment by bartek-wesolowski** [comment]
@rafaeltonholo thank you for such a verbose explanation
